Cpp_test/lib/ipc-shm-send.cpp

```diff
@@ -0,0 +1,159 @@
+#include "ipc-shm-send.h"
+#include <iostream>
+#include <cassert>
+
+IPCShmSend::~IPCShmSend()
+{
+
+	if (this->shm_ptr != nullptr)
+	{
+		if (this->shm_ptr->checkLockMutex == true)
+		{
+			this->shm_ptr->checkLockMutex = false;
+		}
+		int ret1 = pthread_cond_destroy(&(this->shm_ptr->cond_re));
+		if (ret1 != 0)
+		{
+			std::cerr << "Error at pthread_cond_destroy(): receive " << strerror(ret1) << std::endl;
+		}
+		sleep(3);
+		
+		int ret2 = pthread_mutex_trylock(&(this->shm_ptr->mutex));
+		if (ret2 != 0)
+		{
+			std::cerr << "Error at pthread_mutex_trylock(): " << strerror(ret2) << std::endl;
+		}
+		else
+		{
+			int ret3 = pthread_mutex_unlock(&(this->shm_ptr->mutex));
+			if (ret3 != 0)
+			{
+				std::cerr << "Error at pthread_mutex_unlock(): " << strerror(ret3) << std::endl;
+			}
+			else
+			{
+				int ret4 = pthread_mutex_destroy(&(this->shm_ptr->mutex));
+				if (ret4 != 0)
+				{
+					std::cerr << "Error at pthread_mutex_destroy(): " << strerror(ret4) << std::endl;
+				}
+			}
+		}
+
+		int ret5 = pthread_mutexattr_destroy(&(this->mutex_attr));
+		if (ret5 != 0)
+		{
+			std::cerr << "Error at pthread_mutexattr_destroy(): " << strerror(ret5) << std::endl;
+		}
+	}
+}
+
+void IPCShmSend::init()
+{
+	this->file_handler.open_file();
+	this->open_shm();
+	this->set_shm_size();
+	this->map_shm();
+	if (this->shmd > 0)
+	{
+		close(this->shmd);
+	}
+	pthread_mutexattr_destroy(&(this->mutex_attr));
+	this->init_mutex();
+
+	pthread_condattr_destroy(&(this->cond_attr));
+	this->init_cond();
+}
+
+void IPCShmSend::transfer()
+{
+	long read_bytes{0};
+	unsigned long long file_size{this->file_handler.get_file_size()};
+	if (file_size == 0)
+	{
+		throw std::runtime_error("ERROR: File size = 0.");
+	}
+
+	unsigned long long total_sent_bytes{0};
+	std::cout << "Waiting for receiver..." << std::endl;
+	while (!this->shm_ptr->is_end)
+	{
+		if (this->shm_ptr->checkLockMutex == true)
+		{
+			lock_mutex();
+		}
```

Comment:
If think, if the ` checkLockMutex`  is false, you should stop the loop, exit the program and result -1, isn't it?

Reply:
So if it is false, I will throw, because in main I set return -1 when catch throw

Reply:
Yes, exactly, you should implement an exception in case it is false.